lib/html_cleaner.rb

```diff
@@ -290,20 +294,17 @@ def traverse_nodes(node, stack=nil, out_html=nil)
     end
 
     if !node.text? && !node.cdata?
+      out_html += stack.close_and_pop_last(1) if put_outside_p_tag?(node.name) && stack.immediately_inside_paragraph?
```

Comment:
I added this because the HTML5 parser is not happy with structures like `<p><h1>foo</h1><p>`. (Apparently it's stricter about not having these impossible combinations.)